### DISCUSSION TITLE
Concurrent Users Report

### DISCUSSION THREAD

**Discussion by akhiljay** [artilleryio/artillery#1449] [Artillery]
Hi Folks, How do I get access to the concurrent users report? I am trying to load test our websockets server for concurrent users.

**Comment by GeekyDeaks** [comment]
When I last looked at the code, I found that the metric was not exposed.  Simply pre-processing the json before running `artillery report` was pretty trivial though

e.g.

```
const fs = require('fs')

function addInProgress(report) {

    let vusers = 0

    report.intermediate.forEach(i => {
        vusers += i.counters['vusers.created'] || 0
        vusers -= i.counters['vusers.failed'] || 0
        vusers -= i.counters['vusers.completed'] || 0
        i.counters['vusers.inprogress'] = vusers
    })

    return report
}

let data = fs.readFileSync('report.json')
let report = addInProgress(JSON.parse(data))
fs.writeFileSync('report-inprogress.json', JSON.stringify(report, null, 2))
```

For the testing I do, the metric turned out to not be as useful as I originally thought, so I haven't looked any further into it.

**Comment by GeekyDeaks** [comment]
I should probably point out that the above example is a very naive approach as it simply reports the number of vusers still running at the end of the reporting interval, which may not be representative of the concurrent users during that period

**Reply by hassy** [reply]
Defining what "concurrent users" means is tricky, and we don't report it right now for this exact reason... some people want max and some want average, and both of those can depend on how frequently you sample the value during a reporting interval.

**Reply by eliezercazares** [reply]
> Defining what "concurrent users" means is tricky, and we don't report it right now for this exact reason... some people want max and some want average, and both of those can depend on how frequently you sample the value during a reporting interval.

to me it would be the maximum amount of successful communication, as in, send and receive that can be reached before the app crashes. but since WebSockets are fire and forget (I think there is an ack?) we cannot know when the app crashed. Unless the app knows its health and emits alerts before crashing.
